### ISSUE TITLE
why my program run very slower on x86-64 uesd heap profiler?

### ISSUE THREAD

**Issue by alk** [gperftools/gperftools#330]
Originally reported on Google Code with ID 327

```
What steps will reproduce the problem?
1.first,my program run nomal on x86-32 used HEAPPROFILER
2.lasty,i run my program on x86-64 also used HEAPPROFILER.and it run so slowly.
3.please,can it be resolved,how i can do?

What is the expected output? What do you see instead?


What version of the product are you using? On what operating system?


Please provide any additional information below.

```

Reported by `freejuva` on 2011-04-23 17:00:54

**Comment by NickNick** [comment]
I ran into the same issue (very slow heap profiling) and was indeed able to solve it by disabling libunwind (`configure --enable-libunwind=no --enable-framepointers`). Rebuilt my program with -fno-omit-frame-pointers of course and everything worked really well. Happy to have found this!

**Comment by prehistoric-penguin** [comment]
@NickNick 

If you disabled the libunwind usage, it will use GCC's unwind mechanism, it's much slower than libunwind, have you compare d the performance?

**Comment by alk** [comment]
heap profiler is basically "king of slowness". It captures backtraces and other metadata on every call to memory allocator. So easily 3-4 orders of magnitude slower memory allocator performance.

**Comment by albe19029** [comment]
There is a way to resolve perfomance issue borrowed from project https://github.com/KDE/heaptrack/tree/master/src/track
They use unw_backtrace (trace_libunwind.cpp) when it is possible, instead of _Unwind_Backtrace (trace_unwind_tables.cpp)

To make it work with gperftools we must build it with libunwind library installed on build machine. In this case _Unwind_Backtrace will be called.
But if we set TCMALLOC_STACKTRACE_METHOD=generic it will call backtrace.
But if gperftool was build with libunwind, generic will be changed to unw_backtrace (https://man.archlinux.org/man/extra/libunwind/unw_backtrace.3.en)
libunwind weakly aliases backtrace() to unw_backtrace(), so when a program calling backtrace() is linked against libunwind, it may end up calling unw_backtrace().

So finally to solve perfomance issue build gperftools with libunwind and before run set env varialbe TCMALLOC_STACKTRACE_METHOD=generic

**Comment by lindsayad** [comment]
I'm a little bit confused ... based on the documentation my understanding is that without libunwind and with `--enable-frame-pointers` the "builtin" stack unwinder is used. My understanding is that this is supposed to be quite a bit faster. However, @theidexisted mentioned that disabling libunwind leads to using GCC's unwind mechanism, which is actually slower, so I'm a bit confused.

TLDR: what is the fastest configuration for heap profiling? Let's assume for the sake of argument that all my code is compiled with `-fno-omit-frame-pointer`

(Although I am curious to know what happens if my application code is compiled with `-fno-omit-frame-pointer` and my system libraries are not)
